File: detect cycle.py
# Program to Detect Cycle in a Linked List

# Node Class
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

# Function to detect cycle
def detect_cycle(head):
    slow = head
    fast = head

    while fast and fast.next:
        slow = slow.next          # Move 1 step
        fast = fast.next.next     # Move 2 steps

        if slow == fast:
            return True

    return False

# Create Linked List
head = Node(10)
head.next = Node(20)
head.next.next = Node(30)
head.next.next.next = Node(40)

# Create Cycle (40 -> 20)
head.next.next.next.next = head.next

# Check Cycle
if detect_cycle(head):
    print("Cycle Detected")
else:
    print("No Cycle")

File: DetectCycle.java
class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

public class DetectCycle {
    Node head;

    // Detect cycle
    public boolean hasCycle() {
        Node slow = head;
        Node fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;          // 1 step
            fast = fast.next.next;     // 2 steps

            if (slow == fast) {
                return true;
            }
        }

        return false;
    }

    public static void main(String[] args) {
        DetectCycle list = new DetectCycle();

        Node n1 = new Node(10);
        Node n2 = new Node(20);
        Node n3 = new Node(30);
        Node n4 = new Node(40);

        list.head = n1;
        n1.next = n2;
        n2.next = n3;
        n3.next = n4;

        // Create cycle: 40 -> 20
        n4.next = n2;

        if (list.hasCycle()) {
            System.out.println("Cycle Detected");
        } else {
            System.out.println("No Cycle");
        }
    }
}